pkg/storage/registry/spaces/spaces.go

```diff
@@ -0,0 +1,457 @@
+	typesv1beta1 "github.com/cs3org/go-cs3apis/cs3/types/v1beta1"
+	"github.com/cs3org/reva/pkg/appctx"
+	ctxpkg "github.com/cs3org/reva/pkg/ctx"
+	"github.com/cs3org/reva/pkg/errtypes"
+	"github.com/cs3org/reva/pkg/logger"
+	"github.com/cs3org/reva/pkg/rgrpc/status"
+	"github.com/cs3org/reva/pkg/rgrpc/todo/pool"
+	"github.com/cs3org/reva/pkg/sharedconf"
+	"github.com/cs3org/reva/pkg/storage"
+	pkgregistry "github.com/cs3org/reva/pkg/storage/registry/registry"
+	"github.com/mitchellh/mapstructure"
+	"google.golang.org/grpc"
+)
+
+//go:generate mockery -name StorageProviderClient
+
+func init() {
+	pkgregistry.Register("spaces", NewDefault)
+}
+
+type provider struct {
+	Mapping           string            `mapstructure:"mapping"`
+	MountPath         string            `mapstructure:"mount_path"`
+	Aliases           map[string]string `mapstructure:"aliases"`
+	AllowedUserAgents []string          `mapstructure:"allowed_user_agents"`
+	PathTemplate      string            `mapstructure:"path_template"`
+	template          *template.Template
+	// filters
+	SpaceType      string `mapstructure:"space_type"`
+	SpaceOwnerSelf bool   `mapstructure:"space_owner_self"`
+	SpaceID        string `mapstructure:"space_id"`
+}
+
+type templateData struct {
+	CurrentUser *userpb.User
+	Space       *providerpb.StorageSpace
+}
+
+// StorageProviderClient is the interface the spaces registry uses to interact with storage providers
+type StorageProviderClient interface {
+	ListStorageSpaces(ctx context.Context, in *providerpb.ListStorageSpacesRequest, opts ...grpc.CallOption) (*providerpb.ListStorageSpacesResponse, error)
+}
+
+// WithSpace generates a layout based on space data.
+func (p *provider) ProviderPath(u *userpb.User, s *providerpb.StorageSpace) (string, error) {
+	b := bytes.Buffer{}
+	if err := p.template.Execute(&b, templateData{CurrentUser: u, Space: s}); err != nil {
+		return "", err
+	}
+	return b.String(), nil
+}
+
+type config struct {
+	Providers    map[string]*provider `mapstructure:"providers"`
+	HomeTemplate string               `mapstructure:"home_template"`
+}
+
+func (c *config) init() {
+
+	if c.HomeTemplate == "" {
+		c.HomeTemplate = "/"
+	}
+
+	if len(c.Providers) == 0 {
+		c.Providers = map[string]*provider{
+			sharedconf.GetGatewaySVC(""): {
+				MountPath: "/",
+			},
+		}
+	}
+
+	// cleanup rule paths
+	for _, rule := range c.Providers {
+		// if the path template is not explicitly set use the mountpath as path template
+		if rule.PathTemplate == "" && strings.HasPrefix(rule.MountPath, "/") {
+			// TODO err if the path is a regex
+			rule.PathTemplate = rule.MountPath
+		}
+
+		// cleanup path template
+		rule.PathTemplate = filepath.Clean(rule.PathTemplate)
+
+		// compile given template tpl
+		var err error
+		rule.template, err = template.New("path_template").Funcs(sprig.TxtFuncMap()).Parse(rule.PathTemplate)
+		if err != nil {
+			logger.New().Fatal().Err(err).Interface("rule", rule).Msg("error parsing template")
```

Comment:
this is totally worth documenting. In case this service fails, it will cause the entire suite to fail (in the case of ocis running supervised).

Reply:
Done in https://github.com/cs3org/reva/pull/2383